### PR TITLE
enable copyFileToContainer feature during container startup

### PR DESCRIPTION
**Pull request by dharanikesav**
Currently, testcontainer is capable of copying files only after starting a container. But, we have a requirement to copy files to container after creation and then start the container. This code change is to enable copy files to container immediately after creation.

### CODE REVIEW

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 1]
```diff
@@ -940,8 +958,8 @@ public ExecResult execInContainer(String... command)
     @Override
     public void copyFileToContainer(MountableFile mountableLocalFile, String containerPath) {
 
-        if (!isRunning()) {
-            throw new IllegalStateException("copyFileToContainer can only be used while the Container is running");
+        if (!isRunning() && !isCreated()) {
+            throw new IllegalStateException("opyFileToContainer can only be used with created / running container");
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Typo in exception message

File: core/src/main/java/org/testcontainers/containers/Container.java [hunk 2]
```diff
@@ -128,6 +128,14 @@ default void addFileSystemBind(final String hostPath, final String containerPath
      */
     void addExposedPorts(int... ports);
 
+    /**
+     * add a file to be copied before starting a created container
+     *
+     * @param mountableFile a Mountable file with path of source file / folder on host machine
+     * @param containerPath a destination path on conatiner to which the files / folders to be copied
+     */
+    void addCopyFileToContainer(MountableFile mountableFile, String containerPath);
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/Container.java]: We've decided to deprecate the non-fluent setters, so you can omit this method.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 3]
```diff
@@ -835,6 +838,21 @@ public SELF withWorkingDirectory(String workDir) {
         return self();
     }
 
+
+    @Override
+    public void addCopyFileToContainer(MountableFile mountableFile, String containerPath) {
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: See above, you can remove this method.

File: core/out/production/resources/META-INF/services/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+The contents of this folder is a workaround for issues described here: https://github.com/KostyaSha/docker-java-shade/issues/1
```
Comment (bsideup) [inline core/out/production/resources/META-INF/services/README.md]: this file (as well as other `/out` files) should not be commited

File: core/src/main/java/org/testcontainers/containers/ContainerState.java [hunk 5]
```diff
@@ -43,6 +43,22 @@ default boolean isRunning() {
...
+     * @return is the container created?
+     */
+    default boolean isCreated() {
+        if (getContainerId() == null) {
+            return false;
+        }
+
+        try {
+            Boolean created = getCurrentContainerInfo().getState().getStatus().equalsIgnoreCase("Created");
+            return Boolean.TRUE.equals(created);
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/containers/ContainerState.java]: In the `isRunning` method the result of `getRunning()` is a nullable `Boolean`, which is why we do the `TRUE.equals` call in that method.

Here, `equalsIgnoreCase` just returns a primitive boolean, and further, `getStatus()` may return null.

As such, I think it would actually be safer to just replace these two lines with:
`return "created".equalsIgnoreCase(getCurrentContainerInfo().getState().getStatus());`

Please could we do that instead?

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Test
+    public void checkFileCopied() {
+        container.start();
+        try {
+            String filesList = container.execInContainer("ls","/opt/mappable-resource").getStdout();
+
+            Assert.assertTrue(filesList.contains(fileName));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (rnorth) [inline core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: Please let the exception be thrown from the test method, as it could be important.

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+public class CopyFileToContainerTest {
+    private static String folderPath = new File("./src/test/resources/mappable-resource/").getAbsolutePath();
+    private static String containerPath = "/opt";
+    private static String fileName = "test-resource.txt";
+    public static GenericContainer container;
+
+    @Before
+    public void before() {
+        container = new GenericContainer("couchbase:latest")
```
Comment (rnorth) [inline core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: I think we can get away with using a smaller image here (and save some startup time). Would an `alpine` image be OK?

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            .withStartupCheckStrategy(new MinimumDurationRunningStartupCheckStrategy(Duration.ofSeconds(5)))
+            .withCopyFileToContainer(MountableFile.forHostPath(folderPath), containerPath );
+    }
+
+
+    @Test
+    public void checkFileCopied() {
+        container.start();
```
Comment (rnorth) [inline core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: Do we need to split the instantiation/start of the container like this, or could the container just be a field with an `@Rule` annotation for this test? I think I'd prefer the latter, as it's the more idiomatic way.

Comment (bsideup) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: actually, didn't we agree to use `try-with-resources` style? Otherwise, if you have 5 rules in your test and run a single test all of them will start - that sucks :(

File: core/src/main/java/org/testcontainers/containers/ContainerState.java (hunk 5 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/containers/ContainerState.java]: Also, this method will return `false` if the container is not just created but also started

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java (hunk 8 above)
Comment (rnorth) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: Maybe it depends? This class has only one test method (and would have only one Rule), which is why I suggested this way. I'm _very_ relaxed about it though; `try-with-resources` would also be fine.

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java (hunk 7 above)
Comment (dharanikesav) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: we need container to be up and running to perform this test. Alpine container is exiting immediately after start. Because of this, i am unable to use alpine container

Comment (kiview) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: Use the `alpine` image and provide a long running command (like `sleep`), using `withCommand()` on the container.

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java (hunk 8 above)
Comment (dharanikesav) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: What is this try-with-resources? could you please a reference to understand this

Comment (bsideup) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: Sure, see here:
https://github.com/testcontainers/testcontainers-java/blob/609f9a10c3c110be405b9cd96fc340c992b79b3b/core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java#L346

Comment (dharanikesav) [reply core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: @bsideup Thank you.. that helps

**Comment by dharanikesav** [comment]
@kiview @bsideup I have done the changes requested. Is it ok to merge this branch?

File: core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    private static String folderPath = new File("./src/test/resources/mappable-resource/").getAbsolutePath();
+    private static String containerPath = "/tmp";
+    private static String fileName = "test-resource.txt";
+
+    @Test
+    public void checkFileCopied() throws IOException, InterruptedException {
+        try(
+            GenericContainer container = new GenericContainer("alpine:latest")
+                .withCommand("sleep","3000")
+                .withCopyFileToContainer(MountableFile.forHostPath(folderPath), containerPath)
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java]: please use `MountableFile.forClasspathResource("/mappable-resource/test-resource.txt")`

**Comment by dharanikesav** [comment]
@rnorth @bsideup @kiview Incorporated all the changes requested.

File: core/src/test/resources/logback-test.xml [hunk 10]
```diff
@@ -26,4 +26,4 @@
         <Marker>PROFILER</Marker>
         <OnMatch>DENY</OnMatch>
     </turboFilter>
-</configuration>
\ No newline at end of file
+</configuration>
```
Comment (kiview) [inline core/src/test/resources/logback-test.xml]: This is of course an unnecessary change, but rest LGTM.

**Comment by kiview** [comment]
@rnorth Have your remarks been addressed?

**Comment by bsideup** [comment]
@dharanikesav sorry for taking so long for the review! Richard is a bit busy, so I took an action to merge it :)

Thanks for your contribution! A really good one! 👍
